// eslint-plugin-boundaries/dist/Support/Debug.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.warn = warn;
exports.success = success;
exports.warnOnce = warnOnce;
exports.debugDescription = debugDescription;
const elements_1 = require("@boundaries/elements");
const chalk_1 = __importDefault(require("chalk"));
const Settings_1 = require("../Settings");
const warns = [];
const debuggedFiles = [];
const COLORS_MAP = {
    gray: "gray",
    green: "green",
    yellow: "yellow",
};
function trace(message, color) {
    // eslint-disable-next-line no-console
    console.log(chalk_1.default[COLORS_MAP[color]](`[${Settings_1.PLUGIN_NAME}]: ${message}`));
}
function warn(message) {
    trace(message, COLORS_MAP.yellow);
}
function success(message) {
    trace(message, COLORS_MAP.green);
}
function warnOnce(message) {
    if (!warns.includes(message)) {
        warns.push(message);
        warn(message);
    }
}
function debugDescription(elementDescription) {
    const isDependency = (0, elements_1.isDependencyDescription)(elementDescription);
    const key = isDependency
        ? elementDescription.to.source
        : elementDescription.path || "";
    const type = isDependency
        ? elementDescription.to.type
        : elementDescription.type;
    if ((0, Settings_1.isDebugModeEnabled)() && !debuggedFiles.includes(key)) {
        debuggedFiles.push(key);
        if (type) {
            success(`'${key}' is of type '${type}'`);
        }
        else {
            warn(`'${key}' is of unknown type`);
        }
        trace(`\n${JSON.stringify(elementDescription, null, 2)}`, COLORS_MAP.gray);
    }
}
