// eslint-plugin-boundaries/dist/Elements/Elements.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getElementsMatcher = getElementsMatcher;
exports.getSpecifiers = getSpecifiers;
exports.elementDescription = elementDescription;
exports.dependencyDescription = dependencyDescription;
const elements_1 = require("@boundaries/elements");
const resolve_1 = __importDefault(require("eslint-module-utils/resolve"));
const Support_1 = require("../Support");
const elements = new elements_1.Elements();
/**
 * Returns the elements matcher based on the ESLint rule context, filtering out invalid descriptors
 * @param context The ESLint rule context
 * @returns The elements matcher
 */
function getElementsMatcher(settings) {
    const elementsMatcher = elements.getMatcher(settings.elementDescriptors, {
        ignorePaths: settings.ignorePaths,
        includePaths: settings.includePaths,
        legacyTemplates: settings.legacyTemplates,
        cache: settings.cache,
        flagAsExternal: settings.flagAsExternal,
        rootPath: settings.rootPath,
    });
    return elementsMatcher;
}
/**
 * Returns the specifiers used in an import or export statement
 * @param node The AST node representing the import or export
 * @returns The list of specifiers used in the import or export
 */
function getSpecifiers(node) {
    if (node.parent.type === "ImportDeclaration") {
        return node.parent.specifiers
            .filter((specifier) => specifier.type === "ImportSpecifier" &&
            specifier.imported &&
            specifier.imported.name)
            .map((specifier) => specifier.imported.name);
    }
    if (node.parent.type === "ExportNamedDeclaration") {
        return node.parent.specifiers
            .filter((specifier) => specifier.type === "ExportSpecifier" &&
            specifier.exported.name)
            .map((specifier) => specifier.exported.name);
    }
    return [];
}
/**
 * Returns the description of the current file being linted
 * @param fileName The file name (absolute path)
 * @param settings The ESLint rule context settings normalized
 * @returns The description of the current file being linted
 */
function elementDescription(fileName, settings) {
    const matcher = getElementsMatcher(settings);
    const result = matcher.describeElement(fileName);
    (0, Support_1.debugDescription)(result);
    return result;
}
/**
 * Returns the description of a dependency node
 * @param param0 The dependency node info
 * @param context The ESLint rule context
 * @returns The description of the dependency node
 */
function dependencyDescription({ node, kind, nodeKind, }, 
/** The file name (absolute path) */
fileName, 
/** The ESLint rule context settings normalized */
settings, 
/** The ESLint rule context */
context) {
    const source = String(node.value);
    const matcher = getElementsMatcher(settings);
    const resolvedPath = (0, resolve_1.default)(source, context);
    const description = matcher.describeDependency({
        from: fileName,
        to: resolvedPath || undefined,
        source,
        kind: kind || "value", // TODO: Change by runtime in a backwards compatible way
        nodeKind,
        specifiers: getSpecifiers(node),
    });
    (0, Support_1.debugDescription)(description);
    return description;
}
