// eslint-plugin-boundaries/dist/Settings/Helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isImportKind = isImportKind;
exports.isDependencyNodeKey = isDependencyNodeKey;
exports.isSettingsKey = isSettingsKey;
exports.isRulePolicy = isRulePolicy;
exports.isRuleName = isRuleName;
exports.isRuleShortName = isRuleShortName;
exports.isLegacyType = isLegacyType;
exports.rulesMainKey = rulesMainKey;
const elements_1 = require("@boundaries/elements");
const Support_1 = require("../Support");
const Settings_types_1 = require("./Settings.types");
/**
 * Type guard to check if a value is a valid DependencyKind.
 * @param value The value to check.
 * @returns True if the value is a valid DependencyKind, false otherwise.
 * @deprecated Use isDependencyKind instead.
 */
function isImportKind(value) {
    return ((0, Support_1.isString)(value) &&
        Object.values(elements_1.DEPENDENCY_KINDS_MAP).includes(value));
}
/**
 * Type guard to check if a value is a valid DependencyNodeKey.
 * @param value The value to check.
 * @returns True if the value is a valid DependencyNodeKey, false otherwise.
 */
function isDependencyNodeKey(value) {
    return ((0, Support_1.isString)(value) &&
        Object.values(Settings_types_1.DEPENDENCY_NODE_KEYS_MAP).includes(value));
}
/**
 * Type guard to check if a value is a valid key for the plugin settings.
 * @param value - The value to check.
 * @returns True if the value is a valid settings key, false otherwise.
 */
function isSettingsKey(value) {
    return ((0, Support_1.isString)(value) &&
        Object.values(Settings_types_1.SETTINGS_KEYS_MAP).includes(value));
}
/**
 * Type guard to check if a value is a valid RulePolicy.
 * @param value - The value to check.
 * @returns True if the value is a valid RulePolicy, false otherwise.
 */
function isRulePolicy(value) {
    return ((0, Support_1.isString)(value) &&
        (value === Settings_types_1.RULE_POLICY_ALLOW || value === Settings_types_1.RULE_POLICY_DISALLOW));
}
/**
 * Type guard to check if a value is a valid rule name including the default plugin prefix.
 * @param value - The value to check.
 * @returns True if the value is a valid rule name with the default plugin prefix, false otherwise.
 */
function isRuleName(value) {
    return Settings_types_1.RULE_NAMES.includes(value);
}
/**
 * Type guard to check if a value is a valid rule short name.
 * @param value - The value to check.
 * @returns True if the value is a valid rule short name, false otherwise.
 */
function isRuleShortName(value) {
    return Settings_types_1.RULE_SHORT_NAMES.includes(value);
}
function isLegacyType(type) {
    return (0, Support_1.isString)(type);
}
function rulesMainKey(key = Settings_types_1.FROM) {
    return key;
}
