// eslint-plugin-boundaries/dist/Settings/Settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformLegacyTypes = transformLegacyTypes;
exports.getElementsTypeNames = getElementsTypeNames;
exports.getRootPath = getRootPath;
exports.isDebugModeEnabled = isDebugModeEnabled;
const node_path_1 = require("node:path");
const Support_1 = require("../Support");
const Helpers_1 = require("./Helpers");
const Settings_types_1 = require("./Settings.types");
const { VALID_MODES, ROOT_PATH, ENV_ROOT_PATH, DEBUG } = Settings_types_1.SETTINGS;
// TODO, remove in next major version
function transformLegacyTypes(typesFromSettings) {
    const types = typesFromSettings || [];
    if (!(0, Support_1.isArray)(types)) {
        return [];
    }
    return types.map((type) => {
        // backward compatibility with v1
        if ((0, Helpers_1.isLegacyType)(type)) {
            return {
                type: type,
                match: VALID_MODES[0],
                pattern: `${type}/*`,
                capture: ["elementName"],
            };
        }
        // default options
        return {
            match: VALID_MODES[0],
            ...type,
        };
    });
}
function getElementsTypeNames(descriptors) {
    return descriptors.map((element) => element.type).filter(Boolean);
}
function getRootPath(settings) {
    const rootPathUserSetting = process.env[ENV_ROOT_PATH] || settings[ROOT_PATH];
    if (rootPathUserSetting) {
        return (0, node_path_1.isAbsolute)(rootPathUserSetting)
            ? rootPathUserSetting
            : (0, node_path_1.resolve)(process.cwd(), rootPathUserSetting);
    }
    return process.cwd();
}
function isDebugModeEnabled() {
    return process.env[DEBUG];
}
