// eslint-plugin-boundaries/dist/Public/Settings.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSettingsKey = exports.SETTINGS_KEYS_MAP = exports.isDependencyNodeKey = exports.DEPENDENCY_NODE_KEYS_MAP = exports.isImportKind = exports.IMPORT_KINDS_MAP = exports.isElementDescriptorMode = exports.ELEMENT_DESCRIPTOR_MODES_MAP = exports.isDependencyKind = exports.DEPENDENCY_KINDS_MAP = void 0;
var elements_1 = require("@boundaries/elements");
Object.defineProperty(exports, "DEPENDENCY_KINDS_MAP", { enumerable: true, get: function () { return elements_1.DEPENDENCY_KINDS_MAP; } });
Object.defineProperty(exports, "isDependencyKind", { enumerable: true, get: function () { return elements_1.isDependencyKind; } });
const elements_2 = require("@boundaries/elements");
var elements_3 = require("@boundaries/elements");
Object.defineProperty(exports, "ELEMENT_DESCRIPTOR_MODES_MAP", { enumerable: true, get: function () { return elements_3.ELEMENT_DESCRIPTOR_MODES_MAP; } });
Object.defineProperty(exports, "isElementDescriptorMode", { enumerable: true, get: function () { return elements_3.isElementDescriptorMode; } });
/**
 * Map of the kinds of import, either a type import or a value import.
 * @deprecated Use DEPENDENCY_KINDS_MAP instead
 */
exports.IMPORT_KINDS_MAP = elements_2.DEPENDENCY_KINDS_MAP;
var Settings_1 = require("../Settings");
Object.defineProperty(exports, "isImportKind", { enumerable: true, get: function () { return Settings_1.isImportKind; } });
Object.defineProperty(exports, "DEPENDENCY_NODE_KEYS_MAP", { enumerable: true, get: function () { return Settings_1.DEPENDENCY_NODE_KEYS_MAP; } });
Object.defineProperty(exports, "isDependencyNodeKey", { enumerable: true, get: function () { return Settings_1.isDependencyNodeKey; } });
Object.defineProperty(exports, "SETTINGS_KEYS_MAP", { enumerable: true, get: function () { return Settings_1.SETTINGS_KEYS_MAP; } });
Object.defineProperty(exports, "isSettingsKey", { enumerable: true, get: function () { return Settings_1.isSettingsKey; } });
